### PR TITLE
Intern label KVs as unique.Handle[string] 

### PR DESCRIPTION
**Pull request by fasaxc**
## Description

We expect label keys and values to be reused across many endpoints so this saves considerable amounts of storage.  Previously, every JSON map that we decoded would have its own copies of the key and value
strings.

Furthermore, the maps themselves get smaller because each map key/value only has to store an 8-byte handle instead of a 16-byte string. This effect is magnified because maps often have many emptry KV slots, which
now waste only 8 bytes each instead of 16.

Of course, the interning map itself is not free, but it seems to do well against the benchmarks in the calc package.

Before:
```
goos: linux
goarch: amd64
pkg: github.com/projectcalico/calico/felix/calc
cpu: Intel(R) Core(TM) i9-8950HK CPU @ 2.90GHz
BenchmarkSnapshot200Local250kTotal10000TagPols
BenchmarkSnapshot200Local250kTotal10000TagPols-12                     	       1	3390095423 ns/op	       374.5 HeapAllocMB	        29.00 IPSets	     10432 Msgs	     10000 Policies	         3.390 s	947187896 B/op	 9431885 allocs/op
BenchmarkSnapshot200Local10kTotal1000NetsetPols
BenchmarkSnapshot200Local10kTotal1000NetsetPols-12                    	       2	 509773140 ns/op	        81.43 HeapAllocMB	      1000 IPSets	      2403 Msgs	      1000 Policies	         0.4954 s	244513188 B/op	 1851382 allocs/op
BenchmarkSnapshot200Local10kTotal10000NetsetPols
BenchmarkSnapshot200Local10kTotal10000NetsetPols-12                   	       1	5886280527 ns/op	       609.1 HeapAllocMB	     10000 IPSets	     20403 Msgs	     10000 Policies	         5.886 s	2524610176 B/op	17194334 allocs/op
BenchmarkSnapshotThenDeleteLocal200Local250kTotal10000TagPols
BenchmarkSnapshotThenDeleteLocal200Local250kTotal10000TagPols-12      	       1	3786732510 ns/op	       268.8 HeapAllocMB	        29.00 IPSets	     20761 Msgs	     10000 Policies	         3.787 s	1088312568 B/op	 9960248 allocs/op
BenchmarkSnapshotThenDeleteLocal200Local10kTotal1000NetsetPols
BenchmarkSnapshotThenDeleteLocal200Local10kTotal1000NetsetPols-12     	       2	 790657256 ns/op	        22.41 HeapAllocMB	      1000 IPSets	      4703 Msgs	      1000 Policies	         0.7764 s	255755384 B/op	 2119475 allocs/op
BenchmarkSnapshotThenDeleteLocal200Local10kTotal10000NetsetPols
BenchmarkSnapshotThenDeleteLocal200Local10kTotal10000NetsetPols-12    	       1	9490867400 ns/op	        80.39 HeapAllocMB	     10000 IPSets	     40703 Msgs	     10000 Policies	         9.491 s	2629982456 B/op	19795282 allocs/op
PASS
```

After
```
goos: linux
goarch: amd64
pkg: github.com/projectcalico/calico/felix/calc
cpu: Intel(R) Core(TM) i9-8950HK CPU @ 2.90GHz
BenchmarkSnapshot200Local250kTotal10000TagPols
BenchmarkSnapshot200Local250kTotal10000TagPols-12                     	       1	3239373548 ns/op	       307.9 HeapAllocMB	        29.00 IPSets	     10432 Msgs	     10000 Policies	         3.239 s	970191120 B/op	10684881 allocs/op
BenchmarkSnapshot200Local10kTotal1000NetsetPols
BenchmarkSnapshot200Local10kTotal1000NetsetPols-12                    	       2	 507566044 ns/op	        78.60 HeapAllocMB	      1000 IPSets	      2403 Msgs	      1000 Policies	         0.4959 s	245641812 B/op	 1909878 allocs/op
BenchmarkSnapshot200Local10kTotal10000NetsetPols
BenchmarkSnapshot200Local10kTotal10000NetsetPols-12                   	       1	5851975970 ns/op	       604.9 HeapAllocMB	     10000 IPSets	     20403 Msgs	     10000 Policies	         5.852 s	2525118376 B/op	17307165 allocs/op
BenchmarkSnapshotThenDeleteLocal200Local250kTotal10000TagPols
BenchmarkSnapshotThenDeleteLocal200Local250kTotal10000TagPols-12      	       1	4008961719 ns/op	       202.5 HeapAllocMB	        29.00 IPSets	     20761 Msgs	     10000 Policies	         4.009 s	1111335840 B/op	11213888 allocs/op
BenchmarkSnapshotThenDeleteLocal200Local10kTotal1000NetsetPols
BenchmarkSnapshotThenDeleteLocal200Local10kTotal1000NetsetPols-12     	       2	 804245244 ns/op	        19.84 HeapAllocMB	      1000 IPSets	      4703 Msgs	      1000 Policies	         0.7814 s	256900296 B/op	 2179831 allocs/op
BenchmarkSnapshotThenDeleteLocal200Local10kTotal10000NetsetPols
BenchmarkSnapshotThenDeleteLocal200Local10kTotal10000NetsetPols-12    	       1	9498288054 ns/op	        77.71 HeapAllocMB	     10000 IPSets	     40703 Msgs	     10000 Policies	         9.498 s	2630446624 B/op	19918995 allocs/op
PASS
```
## Related issues/PRs

CORE-11313

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
Felix now deduplicates resource label keys and values, significantly reducing memory usage at high scale. 
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: lib/std/uniquelabels/unique_labels.go [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+		return err
+	}
+	i.m = hm
+	return nil
+}
+
+func (i Map) AllHandles() iter.Seq2[uniquestr.Handle, uniquestr.Handle] {
+	return func(yield func(uniquestr.Handle, uniquestr.Handle) bool) {
+		for k, v := range i.m {
+			if !yield(uniquestr.Handle(k), uniquestr.Handle(v)) {
```
Comment (caseydavenport) [inline lib/std/uniquelabels/unique_labels.go:98]: I'm not entirely sure I follow why we need to do a type conversion of `k` and `v` to `uniquestr.Handle` here - aren't these already coming from `i.m`, which is itself a `map[uniquestr.Handle]uniquestr.Handle`?

File: lib/std/uniquelabels/unique_labels_test.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+package uniquelabels
+
+import (
+	"bytes"
+	"encoding/json"
+	"testing"
+)
+
+func TestInternedLabelsJSONRoundTrip(t *testing.T) {
```
Comment (caseydavenport) [inline lib/std/uniquelabels/unique_labels_test.go:23]: I think it's worth filling out this test suite a bit more:

- Additional Equality testcases
- JSON marshal / unmarshal edge cases - e.g., for nil and Empty
- Test cases for `IntersectAndFilter`

File: lib/std/uniquelabels/unique_labels.go [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+// To allow drop-in replacement of a map[string]string, Map has unique
+// representations of nil (Nil) and empty (Empty). The zero value of a Map is
+// the Nil value; it can be detected with IsNil(). Map is a value type, which
+// precludes using Go's normal nil.  Making Map a pointer type would increase
+// overheads; making map an alias for map[uniquestr.Handle]uniquestr.Handle
+// would break the read-only property, preventing future optimisations.
+//
+// Uses uniquestr.Handle internally, so it is most efficient to query the map
+// using AllHandles() and GetHandle().
+type Map struct {
```
Comment (caseydavenport) [inline lib/std/uniquelabels/unique_labels.go:46]: Not really feedback, but for my own understanding I am curious if you considered a single `unique` package with `type Labels` and `type String` instead of separate `uniquelabels` and `uniquestring` packages and what led you in this direction.

**Review by caseydavenport — APPROVED** [review]
This looks pretty sound to me! Mostly just a comment on unit testing and a couple of questions for my own understanding.

File: lib/std/uniquelabels/unique_labels.go (hunk 1 above)
Comment (fasaxc) [reply lib/std/uniquelabels/unique_labels.go:98]: Yep, redundant now, came from a a refactoring

File: lib/std/uniquelabels/unique_labels_test.go (hunk 2 above)
Comment (fasaxc) [reply lib/std/uniquelabels/unique_labels_test.go:23]: Yep, haven't fleshed out the tests yet.

File: lib/std/uniquelabels/unique_labels.go (hunk 3 above)
Comment (fasaxc) [reply lib/std/uniquelabels/unique_labels.go:46]: I tried making `uniquestring` just `unique` but the ambiguity with the stdlib package of the same name was a bit confusing.  I tried a few variations with no super clear winner TBH!  Happy to get a second opinion.

File: lib/std/uniquelabels/unique_labels.go (hunk 1 above)
Comment (fasaxc) [reply lib/std/uniquelabels/unique_labels.go]: Fixed
